src/drivers/virtio/transport/pci.rs

```diff
@@ -702,9 +702,9 @@ impl NotifCfg {
 
 		// Assumes the cap_len is a multiple of 8
 		// This read MIGHT be slow, as it does NOT ensure 32 bit alignment.
-		let notify_off_multiplier = cap.device.read_register(
-			u16::try_from(cap.origin.cfg_ptr).unwrap() + u16::from(cap.origin.cap_struct.cap_len),
-		);
+		let notify_off_multiplier = cap
+			.device
+			.read_register(u16::try_from(cap.origin.cfg_ptr).unwrap() + 16);
```

Comment:
Why do is 16 correct and cap.origin.cap_struct.cap_len wrong? 

Is the following line also wrong?

https://github.com/hermit-os/kernel/blob/main/src/drivers/virtio/transport/pci.rs#L950-L952

Reply:
This needs to read the last 4 bytes of the struct. Adding `cap_len` results in reading the first byte of the next struct. Likely that the other calculation is also incorrect.

Reply:
I have fixed the other two uses of `cfg_ptr` too in the latest commits. Could you take a look?

Reply:
A little verbose but looks good :+1: